sdks/java/core/src/main/java/org/apache/beam/sdk/coders/AvroCoder.java

```diff
@@ -696,16 +696,17 @@ private void checkArray(String context, TypeDescriptor<?> type, Schema schema) {
      * see private fields. We may need to walk up to the parent to get classes from the parent.
      */
     private static Field getField(Class<?> clazz, String name) {
-      while (clazz != null) {
-        for (Field field : clazz.getDeclaredFields()) {
+      Class<?> currentClass = clazz;
+      while (currentClass != null) {
+        for (Field field : currentClass.getDeclaredFields()) {
           AvroName avroName = field.getAnnotation(AvroName.class);
           if (avroName != null && name.equals(avroName.value())) {
             return field;
           } else if (avroName == null && name.equals(field.getName())) {
             return field;
           }
         }
-        clazz = clazz.getSuperclass();
+        currentClass = currentClass.getSuperclass();
       }
 
       throw new IllegalArgumentException(
```

Comment:
FYI, bug was here: previously we were iterating through values of clazz, and would only reach this error condition if the value was null. We should use the input value rather than the iteration value.